enhancements/microshift/microshift-default-cni.md

```diff
@@ -0,0 +1,290 @@
+
+An application developer builds applications where:
+1. Pods need to talk to each other via TCP o UDP connectivity.
+2. Pods need to talk to services on the LAN or Internet via TCP or UDP.
+3. Some pods need be exposed as NodePort types of services on the LAN.
+4. Some pods need to be exposed to other pods as ClusterIP services types.
+
+An application developer builds applications where some pods must
+have limited connectivity to other pods, or to the external network (i.e.
+plugin services from vendors which must only contact specific internal APIs)
+
+The edge device with MicroShift and the final application is deployed on an
+IPv6, or DualStack IPV4/IPV6 network.
+
+As a device owner I can protect the edge device with a host-level firewall
+(firewalld, iptables, nftables), and that type of configuration is compatible
+with the CNI.
+
+The edge device should be capable of deploying on a network with dynamic address
+provisioning like DHCP, SLAAC, or DHCPv6 and remain functional when the IP address
+changes.
+
+### Goals
+ 
+* Implement a default CNI that satisfies the described user histories while
+  minimizing the impact on footprint.
+
+* The CNI plugin should run without the ClusterNetworkOperator, providing
+  the `microshift-networking` elements any basic system configuration elements,
+  tools or dependencies to make the CNI plugin work.
+
+* Follow the [MicroShift design
+  principles](https://github.com/openshift/microshift/blob/main/docs/design.md)
+
+### Non-Goals
+
+* We do not plan to support multi-node clusters.
+* We are not trying to address high availability within the CNI, 
+  high availability should be acomplished by having multiple edge devices working
+  as single node MicroShifts.
+
+* LoadBalancer type of Services, as this is out of the scope of the CNI.
+
+## Proposal
+
+To meet the requirements described in the user histories, as well footprint
+requirements of the edge environment, an optimized version of OVN Kubernetes
+is proposed.
+
+This option aligns better with Red Hat's strategy on networking and avoids
+the need to maintain bridge-cni which is deprecated in podman 4.
+
+Using OVN-kubernetes is only possible thanks to the following optimizations:
+* Single node operation
+* Communication via unix sockets, avoiding the need of TLS & Certificate creation
+  at boot time
+* No leader election, as there is no high availability or multi-node operation.
+* OpenvSwitch and ovsdb-server RAM optimization (CPUAffinity=0 and --no-mlockall)
+* Workload partitioning, to limit the cpusets where specific ovn-kubernetes workloads
+  can run.
+* New ovn-kubernetes with reduced footprint, oriented to single node operation.
+
+The optimizations reduce the total footprint impact of OVN-Kubernetes + OpenvSwitch
+down to:
+* 100MB of RAM (from +600MB)
+* 366MB of Disk (from 1GB)
+* 0.011CPU cores (based on an i7-7600 platform or a Xeon 4210 VM with 1 VCPU)
+
+Remaining work items in ovn-kubernetes:
+* Handling node IP changes
```

Comment:
There will be other optimizations as we apply the single-node rule to various ovnk features and dig into shared libraries and golang internals, this will be a continuous work. it might be good to mention optimizations will not be restricted to the above.

Reply:
For reference, some ideas for future improvement:
```
Ovnk code efficiency
- Use index, not list all. I.e.  getInterface(index) vs getInterfaces()
- Avoid allocate larger buffer than what is needed
- Use less informers or event broadcaster etc where possible

Apply single node operation rule to ovnk
- Disable ingress ip rotation
- Disable ovnk leader election
- Ovn databases
- Simplified ovn topology?

Golang library efficiency
- cgo enable/disable comparison
- Libovsdb, netlink libraries
- Analyse heap memory usage via pprof

Debugging improvement
- “-s -w” build flag to remove debug info which in turn reduces binary size
- only enable pprof while debugging
- simplify network must-gather
- Move debugging tools from ovnk image to network tools

Metrics
- Fully disable metrics when not needed.

Golang internal related
- Field alignment
- Efficient garbage collection

Ovs & ovn
```

Reply:
If we have a list of planned future work, it is good to include it. Please make sure it's clear which changes will be made for sure and which are investigation we are considering for the future.

Reply:
Are we ready to resolve this thread?

Reply:
Yes, it is done.
@dhellmann Could you help to resolve the comment? I don't have write access to the repo.